Dependency scraper

dbluhm opened the pull request:
This PR adds a maven plugin that can be to retrieve files from dependencies and copy them to another location in the current project. This is primarily intended for use in Vaadin projects that use Data Elements from dependencies and being able to pull TypeScript representations of those data elements and other auto-generated TypeScript classes into the projects frontend sources.

This PR formerly included basic TypeScript generation but I removed it with the intent of keeping the change set limited for this PR.

jayjaybillings commented on org.eclipse.ice.dev/org.eclipse.ice.dev.dependencyscraper/src/main/java/org/org/eclipse/ice/dev/dependencyscraper/DependencyScraper.java:
```diff
@@ -0,0 +1,227 @@
+				.indexOf(sourceDirectory.toLowerCase(Locale.ENGLISH))
+				+ sourceDirectory.length());
+		// Determine output path
+		File target = new File(outputDirectory, relativePath);
+		try {
+			if (target.exists()) {
+				if (clobber) { // Overwrite existing file
+					getLog().info(String.format(
+						"File %s already exists and clobber is set; overwriting.",
+						target.getName()
+					));
+
+					// Replace target only if contents differ
+					File tempFile = File.createTempFile(fullPath, null);
+					FileUtils.copyInputStreamToFile(
+						jar.getInputStream(fileInJar), tempFile
+					);
+					if (!FileUtils.contentEquals(tempFile, target)) {
+						FileUtils.forceDelete(target);
+						FileUtils.moveFile(tempFile, target);
+					} else {
+						tempFile.delete();
+					}
+				} else {
+					getLog().info(String.format(
+						"File %s already exists and clobber is not set; skipping.",
+						target.getName()
+					));
+				}
+			} else {
+				FileUtils.copyInputStreamToFile(
+					jar.getInputStream(fileInJar),
+					target
+				);
+			}
```
I would prefer to break this into at least two functions because there are four layers of branching.

jayjaybillings commented on org.eclipse.ice.dev/org.eclipse.ice.dev.dependencyscraper/src/main/java/org/org/eclipse/ice/dev/dependencyscraper/DependencyScraper.java:
```diff
@@ -0,0 +1,227 @@
+	 * directory specified by parameters.
+	 */
+	public void execute() throws MojoExecutionException, MojoFailureException {
+		if (jarFiles == null) {
+			this.jarFiles = project.getArtifacts().stream()
+				.filter(artifact -> "jar".equals(artifact.getType()))
+				.map(Artifact::getFile).collect(Collectors.toSet());
+		}
+		for (File jar : jarFiles) {
+			try (JarFile jarFile = new JarFile(jar, false)) {
+				Set<ZipEntry> toCopy = jarFile.stream()
+					.filter(file -> !file.isDirectory())
+					.filter(this::startsWithSourceDirectory)
+					.filter(this::shouldInclude)
+					.collect(Collectors.toSet());
+				for (ZipEntry file : toCopy) {
+					getLog().info(String.format("Copying %s", file.getName()));
+					copyFileInJarToOutput(jarFile, file);
+				}
+			} catch (IOException e) {
+				throw new MojoFailureException(
+					String.format("Failed to open jar file %s!", jar.toString())
+				);
+			}
+		}
```
More documentation would be nice, even if only a few lines. All of the big functional blocks, like 205-207 and 211-214 should have at least a few lines describing what they are doing because functional code is inherently confusing. Personally I would put one line per filter in addition a one line initial description.
```java
// Get the set of files to copy
Set<ZipEntry> toCopy = jarFile.stream()
   // No directories
   .filter(file -> !file.isDirectory())
   // Is in the source directory
   .filter(this::startsWithSourceDirectory)
   // Matches the inclusion filter for this type
  .filter(this::shouldInclude).collect(Collectors.toSet());
```

jayjaybillings commented on org.eclipse.ice.dev/org.eclipse.ice.dev.dependencyscraper/src/test/java/org/org/eclipse/ice/dev/dependencyscraper/DependencyScraperTest.java:
```diff
@@ -0,0 +1,193 @@
+import java.util.List;
+import java.util.Set;
+
+public class DependencyScraperTest
```
Documentation please

jayjaybillings commented on org.eclipse.ice.dev/org.eclipse.ice.dev.dependencyscraper/src/test/java/org/org/eclipse/ice/dev/dependencyscraper/DependencyScraperTest.java:
```diff
@@ -0,0 +1,193 @@
+	private static final Path PROJECT = Path.of("target/test-classes/project-to-test");
+	private static final Path OUTPUT = Path.of("target/test-classes/project-to-test/test");
+
+	@Rule
```
What's the rule do? (i.e. - Docs please!)

jayjaybillings reviewed (changes requested):
What's the src/it folder for? Are we now using Groovy as a dependency?

Your test looks comprehensive. What's your coverage level?

dbluhm commented:
> What's the src/it folder for? Are we now using Groovy as a dependency?

Ah, woops, that was included in the archetype I used to generate the base project. Fortunately, the plugin is simple enough that I don't think we'll require integration tests so I will delete that and the associated pieces. Thanks for catching that.

> Your test looks comprehensive. What's your coverage level?

88.5% which is basically everything but a couple of `IOException` handlers. Initially I wasn't sure I'd be able to get it to hit those handlers in testing but looking at it again I think I could probably hit at least one scenario that would throw an IO Exception.

dbluhm commented:
Addressed feedback and increased coverage to 98% :slightly_smiling_face:

jayjaybillings commented:
Nice work on the coverage! :) Is there any reason that this is not merged other than me? I think you addressed my questions, correct?

dbluhm commented:
> Nice work on the coverage! :) Is there any reason that this is not merged other than me? I think you addressed my questions, correct?

I think that's everything! Just waiting to make sure I'd adequately addressed your comments before merging

jayjaybillings reviewed (approved):
Nice work.
